File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlInsetIcon.kt
package com.dizyx.nockerl.design.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import com.dizyx.nockerl.design.tokens.nockerlInsetTaper

/**
 * The **inset icon** is the ratified INFORMATIONAL icon treatment ( canon,
 * law ): a glyph that SINKS into a recessed disc well ("fields sink") rather
 * than sitting on a raised filled-circle. the design lead's "super classy" direction,
 * first shipped on the EmptyState mark and now the canon for any purely
 * informational icon (EmptyState, Banner marks, panel-header cogs).
 *
 * **NEVER clickable (law ).** Flat/plain + filled-circle icons signal
 * *interactive*; the inset treatment signals *informational*. This composable
 * deliberately exposes NO click handler. An inset icon with an `onClick` is a
 * law violation by construction, so the type makes it impossible.
 *
 * The well is the recessed grammar: circular, with a MUTED glyph on `canvasAlt`
 * + a hairline. The inner shade FOLLOWS the circle ([nockerlInsetTaper], ): a
 * top-weighted radial that wraps the arc + a bottom catch-light, not an abrupt band. Three tones mirror
 * the web `.nk-es__well` variants exactly:
 * - [NockerlInsetIconTone.NEUTRAL]: muted `onCardMuted` mark on the plain well.
 * - [NockerlInsetIconTone.BRAND] is earned cyan: `accentPrimary` on `accentPrimarySoft`.
 * - [NockerlInsetIconTone.ERROR] is warm status: `statusError` on a 12% wash.
 *
 * @param icon the informational glyph.
 * @param contentDescription the a11y name, or `null` when a sibling label already
 *   carries the meaning (e.g. an EmptyState title), then the mark is decorative.
 * @param modifier outer modifier.
 * @param tone the well tint (default neutral).
 */
@Composable
fun NockerlInsetIcon(
    icon: ImageVector,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    tone: NockerlInsetIconTone = NockerlInsetIconTone.NEUTRAL,
) {
    val colors = LocalNockerlColors.current
    val glyph: Color
    val fill: Color
    val border: Color
    when (tone) {
        NockerlInsetIconTone.NEUTRAL -> {
            glyph = colors.onCardMuted
            fill = colors.canvasAlt
            border = colors.cardHairline
        }
        NockerlInsetIconTone.BRAND -> {
            glyph = colors.accentPrimary
            fill = colors.accentPrimarySoft
            border = colors.accentPrimary.copy(alpha = BRAND_BORDER_ALPHA)
        }
        NockerlInsetIconTone.ERROR -> {
            glyph = colors.statusError
            fill = colors.statusError.copy(alpha = ERROR_FILL_ALPHA)
            border = colors.statusError.copy(alpha = ERROR_BORDER_ALPHA)
        }
    }

    Box(
        modifier =
            modifier
                .size(WELL_SIZE)
                .nockerlInsetTaper(shape = CircleShape)
                .background(fill, CircleShape)
                .border(1.dp, border, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector = icon,
            contentDescription = contentDescription,
            tint = glyph,
            modifier = Modifier.size(GLYPH_SIZE),
        )
    }
}

/** The informational inset-icon tones, mirroring the web `.nk-es__well` variants. */
enum class NockerlInsetIconTone {
    /** Muted neutral mark: the default informational look. */
    NEUTRAL,

    /** Earned cyan: first-run / brand-positive empty states. */
    BRAND,

    /** Warm status-error: failure / problem states (never color alone). */
    ERROR,
}

/** Well diameter: the web `space-16` (64). */
private val WELL_SIZE: Dp = 64.dp

/** Glyph diameter inside the well: the web `space-8` (32). */
private val GLYPH_SIZE: Dp = 32.dp

/** Brand border alpha (web `accentPrimary 24%`). */
private const val BRAND_BORDER_ALPHA = 0.24f

/** Error fill wash alpha (web `statusError 12%`). */
private const val ERROR_FILL_ALPHA = 0.12f

/** Error border alpha (web `statusError 30%`). */
private const val ERROR_BORDER_ALPHA = 0.30f

File: Sources/NockerlDesign/Components/NockerlInsetIcon.swift
// NockerlInsetIcon: the ratified INFORMATIONAL icon treatment. A glyph that
// SINKS into a recessed disc well ("fields sink") rather than
// sitting on a raised filled-circle, per the design lead's "super classy" direction, first on
// the EmptyState mark and now canon for any purely informational icon.
//
// NEVER clickable: flat/plain + filled-circle icons signal interactive;
// the inset treatment signals informational. This view exposes NO action. An inset
// icon with a tap handler is a law violation by construction, so the type forbids it.
//
// The well is the recessed grammar shared with the callout/checkbox (canvasAlt +
// inner top shade + hairline), circular, with a MUTED SF Symbol. Three tones mirror
// the web `.nk-es__well` variants 1:1. Mirrors the Compose NockerlInsetIcon.

import SwiftUI

/// The informational inset-icon tones, mirroring the web `.nk-es__well` variants.
public enum NockerlInsetIconTone: Equatable {
    /// Muted neutral mark, the default informational look.
    case neutral
    /// Earned cyan: first-run / brand-positive states.
    case brand
    /// Warm status-error: failure / problem states (never color alone).
    case error
}

/// Inset-icon size. `.md` is the empty-state mark (64pt well / 30pt glyph); `.sm` is an
/// inline half-size (32 / 15) for status rows next to text. Mirror to web/Compose when
/// they add sizes.
public enum NockerlInsetIconSize: Equatable {
    case sm, md
    var well: CGFloat { self == .sm ? 32 : 64 }
    var glyph: CGFloat { self == .sm ? 15 : 30 }
}

/// A purely-informational icon in a recessed well (see the file header). No tap
/// affordance by design.
public struct NockerlInsetIcon: View {
    private let systemName: String
    private let accessibilityLabel: String?
    private let tone: NockerlInsetIconTone
    private let size: NockerlInsetIconSize

    /// Well diameter, the web `space-16` (64). Default (`.md`).
    static let wellSize: CGFloat = 64
    /// Glyph point size inside the well (~the web 32px mark). Default (`.md`).
    static let glyphSize: CGFloat = 30
    /// Brand border opacity (web `accentPrimary 24%`).
    static let brandBorderAlpha: Double = 0.24
    /// Error fill wash opacity (web `statusError 12%`).
    static let errorFillAlpha: Double = 0.12
    /// Error border opacity (web `statusError 30%`).
    static let errorBorderAlpha: Double = 0.30

    /// Create an inset icon.
    /// - Parameters:
    ///   - systemName: the SF Symbol name.
    ///   - accessibilityLabel: the a11y name, or `nil` when a sibling label already
    ///     carries the meaning (then the mark is decorative / hidden).
    ///   - tone: the well tint (default neutral).
    ///   - size: `.md` (64) empty-state, or `.sm` (32) inline. Default `.md`.
    public init(
        systemName: String,
        accessibilityLabel: String? = nil,
        tone: NockerlInsetIconTone = .neutral,
        size: NockerlInsetIconSize = .md
    ) {
        self.systemName = systemName
        self.accessibilityLabel = accessibilityLabel
        self.tone = tone
        self.size = size
    }

    public var body: some View {
        InsetBody(systemName: systemName, accessibilityLabel: accessibilityLabel, tone: tone, size: size)
    }

    /// Environment-reading render body.
    private struct InsetBody: View {
        let systemName: String
        let accessibilityLabel: String?
        let tone: NockerlInsetIconTone
        let size: NockerlInsetIconSize

        @Environment(\.colorScheme) private var colorScheme

        var body: some View {
            let palette = NockerlPalette.resolve(colorScheme)
            let glyph = glyphColor(palette)
            let fill = fillColor(palette)
            let border = borderColor(palette)

            Image(systemName: systemName)
                .font(.system(size: size.glyph, weight: .regular))
                .foregroundColor(glyph)
                //  recessed-disc chrome, via the ONE shared primitive (v1.12.1 unify:
                // the recipe no longer lives inline, so it can't fork from StatCard's disc).
                .nockerlRecessedDisc(fill: fill, border: border, size: size.well)
                // A named inset icon reads as an image; a nameless one is
                // decorative (the sibling label carries meaning) and is hidden.
                .accessibilityHidden(accessibilityLabel == nil)
                .accessibilityLabel(accessibilityLabel ?? "")
        }

        private func glyphColor(_ p: NockerlPalette) -> Color {
            switch tone {
            case .neutral: return p.onCardMuted
            case .brand: return p.accentPrimary
            case .error: return p.statusError
            }
        }

        private func fillColor(_ p: NockerlPalette) -> Color {
            switch tone {
            case .neutral: return p.canvasAlt
            // v1.12.1: TONED cyan via the shared brand fill (was `accentPrimarySoft`).
            case .brand: return NockerlRecessedDiscTone.brandFill(p)
            case .error: return p.statusError.opacity(NockerlInsetIcon.errorFillAlpha)
            }
        }

        private func borderColor(_ p: NockerlPalette) -> Color {
            switch tone {
            case .neutral: return p.cardHairline
            case .brand: return p.accentPrimary.opacity(NockerlInsetIcon.brandBorderAlpha)
            case .error: return p.statusError.opacity(NockerlInsetIcon.errorBorderAlpha)
            }
        }
    }

}
